web/app/help/operations/compute-modularity-of-segmentation.asciidoc

```diff
@@ -0,0 +1,25 @@
+### Compute modularity of segmentation
```

Comment:
We already have a backend operation that computes this and we expose its result when we use find modular clustering. (https://github.com/lynxkite/lynxkite/blob/master/app/com/lynxanalytics/biggraph/frontend_operations/BuildSegmentationOperations.scala#L254) This is a bit suboptimal. But I guess we can leave with it if at least the results are the same. But it would be very bad if computing the modularity with this operation on the result of a find modular clustering op yielded a different number. Can you check this? Also, we can consider using this operation instead of the old one in that FE operation. Somewhat related, doesn't networkit has a better find modular clustering implementation? Maybe we should replace that, too?

Reply:
> We already have a backend operation that computes this and we expose its result when we use find modular clustering.

Oh I didn't realize that! Why didn't we expose it as a box...?

> This is a bit suboptimal.

We have Sphynx implementations of a bunch of Spark operations. What is not to like?

> But I guess we can leave with it if at least the results are the same. But it would be very bad if computing the modularity with this operation on the result of a find modular clustering op yielded a different number. Can you check this?

![image](https://user-images.githubusercontent.com/1268018/104006736-2fe72d80-51a7-11eb-8c11-c863af35a909.png)

Is that close enough? 😸 

Looking at the [NetworKit](https://github.com/networkit/networkit/blob/8.0/networkit/cpp/community/Modularity.cpp) and [LynxKite](https://github.com/lynxkite/lynxkite/blob/4.1.0/app/com/lynxanalytics/biggraph/graph_operations/Modularity.scala) code, I thought the only difference is that we consider the in and out degrees separately, while they just look at the total degree and assume it's 50/50 in/out. But even when I change our code to do the same, the results don't match. I guess I'll debug this a bit.

> Somewhat related, doesn't networkit has a better find modular clustering implementation? Maybe we should replace that, too?

Your implementation is based on the Louvain method, right? That is already integrated as the "Find communities with the Louvain method" box. I don't know if it's compatible enough to be a replacement.

Reply:
> We have Sphynx implementations of a bunch of Spark operations. What is not to like?
Yes, but normally they are just different implementations of the same abstract operation that supposed to do the same thing.

> Your implementation is based on the Louvain method, right?
Well, I would think not, at least not exactly. It has some not too pretty stuff to do things in a distributed way and there is no hierarchical aspect in my implementation.

Reply:
> Looking at the [NetworKit](https://github.com/networkit/networkit/blob/8.0/networkit/cpp/community/Modularity.cpp) and [LynxKite](https://github.com/lynxkite/lynxkite/blob/4.1.0/app/com/lynxanalytics/biggraph/graph_operations/Modularity.scala) code, I thought the only difference is that we consider the in and out degrees separately, while they just look at the total degree and assume it's 50/50 in/out. But even when I change our code to do the same, the results don't match. I guess I'll debug this a bit.

NetworKit just implements Modularity for undirected graphs. An equivalent directed graph contains for every undirected edge `{u, v}` two edges `(u, v)` and `(v, u)`. Thus, the undirected degree is the same as the in-degree and the out-degree and the number of edges in the directed graph is twice the number of edges in the undirected graph. I hope that helps with debugging. I would be surprised if the NetworKit version was wrong, but I of course cannot fully exclude that.

Regarding distributed modularity-based clustering, are you aware of our work at https://github.com/kit-algo/distributed_clustering_thrill? It is implemented in thrill, but from the few source files I have seen so far, the algorithmic ideas should be directly translatable into your framework. If you want high quality, I suggest having a look at the Leiden algorithm, an extension of the Louvain algorithm (see, e.g., https://github.com/vtraag/leidenalg). With some luck (I have a student working on it), we might get an implementation of Leiden in NetworKit till the end of February.

Reply:
> > Looking at the [NetworKit](https://github.com/networkit/networkit/blob/8.0/networkit/cpp/community/Modularity.cpp) and [LynxKite](https://github.com/lynxkite/lynxkite/blob/4.1.0/app/com/lynxanalytics/biggraph/graph_operations/Modularity.scala) code, I thought the only difference is that we consider the in and out degrees separately, while they just look at the total degree and assume it's 50/50 in/out. But even when I change our code to do the same, the results don't match. I guess I'll debug this a bit.
> 
> NetworKit just implements Modularity for undirected graphs.

Thanks! You're right, when I give it an undirected graph I get the same number as in LynxKite.

@xandrew-lynx: So now we have two ways to compute the same scalar. I prefer the new API because it made it easy to add all the metrics the same way. And I don't mind leaving the old LynxKite Modularity op as it is. It's only used in "Find modular clustering", which is in Spark anyway — I don't think we would gain much by moving the whole graph to Sphynx just for this scalar.

> Regarding distributed modularity-based clustering, are you aware of our work at https://github.com/kit-algo/distributed_clustering_thrill? It is implemented in thrill, but from the few source files I have seen so far, the algorithmic ideas should be directly translatable into your framework. If you want high quality, I suggest having a look at the Leiden algorithm, an extension of the Louvain algorithm (see, e.g., https://github.com/vtraag/leidenalg). With some luck (I have a student working on it), we might get an implementation of Leiden in NetworKit till the end of February.

Thanks for the links! These are all fantastic!